### PR TITLE
Initial attempt to persist InvocationTransaction

### PR DESCRIPTION
**Pull request by osmirnov**
@shargon I am trying to implement InvocationTransaction persistence. I attempted to hook up the VM. The execution always returns false as a result. Probably, because I need to register APIs. Please verify my brainstorm dump how it works :D #521

### CODE REVIEW

**Comment by shargon** [comment]
Do you have Windows or Linux?

**Comment by shargon** [comment]
`GetMessage` on `IMessageProvider` should return the TX content, so we can't reuse the same `ExecutionEngineArgs` because the signature of the TX is checked here 

https://github.com/neo-project/neo-vm/blob/master/src/neo-vm/ExecutionEngine.cs#L590

**Comment by osmirnov** [comment]
I am on Windows. We use yours HyperVM not NeoVM. How is your comment applicable?

**Comment by shargon** [comment]
Maybe we have to make `Transaction` and `Block` inherit from `IMessageProvider` but for check witness this is necessary, we need to pass the current "Message"

**Comment by osmirnov** [comment]
Now I stuck with "FaultByGas" error. HyperVM thinks I owe some Gas to it.

image

**Comment by belane** [comment]
Perhaps because is not taking in account that transactions have 10 free gas execution, most of the invokations in the blockchain are bellow 10 of gas.

If you call `engine.Execute((uint)transaction.Gas.Value))` with zero, HyperVM will run out of gas and returns FAULT_BY_GAS.

So it's running https://github.com/CityOfZion/neo-hypervm/blob/master/src/Neo.HyperVM/ExecutionEngine.cpp#L115-L117 with 0 
and many OpCodes add Gas execution (https://github.com/CityOfZion/neo-hypervm/blob/master/src/Neo.HyperVM/ExecutionEngine.cpp#L317) and 
will break contract execution 
https://github.com/CityOfZion/neo-hypervm/blob/master/src/Neo.HyperVM/ExecutionEngine.h#L72


Would you try to run transactions whit less than 10 gas with a default 10 `engine.Execute(10u)` ?

**Comment by osmirnov** [comment]
It is not my transaction. I've got them from blockchain during sync. I expect the gas they have should reflect consumed gas as they are in historical blocks.

File: src/NeoSharp.Core/SmartContract/ScriptTable.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+        public ScriptTable(IRepository repository)
+        {
+            _repository = repository;
+        }
+
+        public byte[] GetScript(byte[] scriptHash, bool isDynamicInvoke)
+        {
+            var contractHash = new UInt160(scriptHash);
+            var contract = _repository.GetContract(contractHash).Result;
```
Comment (shargon) [inline src/NeoSharp.Core/SmartContract/ScriptTable.cs:19]: should check dynamicInvoke param

**Review by shargon — APPROVED** [review]
i can finish it in a week

**Comment by osmirnov** [comment]
@shargon It would be nice if you integrated HyperVM. I currently work on NeoVM integration and my secondary goal to making VMs swappable. So you finish this part I can integrate both VMs faster.

**Pull request merged**